File: admin/src/api/erp_receipt.ts
import { PaginatedRequest, PaginatedResponse } from "@/base/page";
import { api } from "@/utils/request";
import { ErpReceiptDetailBaseResponse, ErpReceiptDetailRequest } from "./erp_receipt_detail";
import { ErpReceiptAttachmentBaseResponse, ErpReceiptAttachmentRequest } from "./erp_receipt_attachment";

const apis = {
  create: "/erp/erp_receipt/create", // 新增
  update: "/erp/erp_receipt/update", // 修改
  delete: "/erp/erp_receipt/delete", // 删除
  get: "/erp/erp_receipt/get", // 单条查询
  get_base: "/erp/erp_receipt/get_base", // 单条查询
  get_info: "/erp/erp_receipt/get_info", // 单条查询
  list: "/erp/erp_receipt/list", // 列表查询
  page: "/erp/erp_receipt/page", // 分页查询
};

export interface ErpReceiptRequest {
  id?: number; // 收款ID
  customer_id: number; // 客户ID
  settlement_account_id: number; // 结算账户ID
  amount: number; // 收款金额
  discount_amount: number; // 优惠金额
  receipt_date: string; // 收款日期
  payment_method: string; // 收款方式 (如 bank_transfer, cash, credit)
  remarks: string; // 备注

  details: ErpReceiptDetailRequest[]; // 收款列表
  attachments: ErpReceiptAttachmentRequest[]; // 收款附件列表
}

export interface ErpReceiptResponse {
  id: number; // 收款ID
  order_number: number; // 订单编号
  customer_id: number; // 客户ID
  user_id: number; // 关联用户ID
  settlement_account_id: number; // 结算账户ID
  amount: number; // 收款金额
  discount_amount: number; // 优惠金额
  receipt_date: string; // 收款日期
  payment_method: string; // 收款方式 (如 bank_transfer, cash, credit)
  receipt_status: number; // 状态 (0=pending, 1=completed, 2=cancelled)
  remarks: string; // 备注
  department_code: string; // 部门编码
  department_id: number; // 部门ID
  creator: number; // 创建者ID
  create_time: string; // 创建时间
  updater: number; // 更新者ID
  update_time: string; // 更新时间

  customer_name: string; // 客户名
  settlement_account_name: string; // 结算账户名

  details: ErpReceiptDetailBaseResponse[]; // 收款列表
  attachments: ErpReceiptAttachmentBaseResponse[]; // 收款附件列表
}

export interface ErpReceiptQueryCondition extends PaginatedRequest {}

export const createErpReceipt = (
  erp_receipt: ErpReceiptRequest
): Promise<number> => {
  return api.post<number>(apis.create, erp_receipt);
};

export const updateErpReceipt = (
  erp_receipt: ErpReceiptRequest
): Promise<void> => {
  return api.post<void>(apis.update, erp_receipt);
};

export const deleteErpReceipt = (id: number): Promise<void> => {
  return api.post<void>(`${apis.delete}/${id}`);
};

export const getErpReceipt = (id: number): Promise<ErpReceiptResponse> => {
  return api.get<ErpReceiptResponse>(`${apis.get}/${id}`);
};

export const getBaseErpReceipt = (id: number): Promise<ErpReceiptResponse> => {
  return api.get<ErpReceiptResponse>(`${apis.get_base}/${id}`);
};

export const getInfoErpReceipt = (id: number): Promise<ErpReceiptResponse> => {
  return api.get<ErpReceiptResponse>(`${apis.get_info}/${id}`);
};

export const listErpReceipt = (): Promise<Array<ErpReceiptResponse>> => {
  return api.get<Array<ErpReceiptResponse>>(apis.list);
};

export const pageErpReceipt = (
  condition: ErpReceiptQueryCondition
): Promise<PaginatedResponse<ErpReceiptResponse>> => {
  return api.get<PaginatedResponse<ErpReceiptResponse>>(apis.page, condition);
};

File: server/module/module-erp/erp-model/src/response/erp_receipt.rs
use chrono::NaiveDateTime;
use serde::{Serialize, Deserialize};
use utoipa::ToSchema;
use serde_with::{serde_as, DisplayFromStr};
use common::formatter::string_date_time::StringDateTime;

use crate::response::{erp_receipt_attachment::ErpReceiptAttachmentBaseResponse, erp_receipt_detail::ErpReceiptDetailBaseResponse};

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpReceiptResponse {
    
    pub id: i64, // 收款ID
    
    pub customer_id: i64, // 客户ID
    
    pub user_id: i64, // 关联用户ID
    
    pub settlement_account_id: Option<i64>, // 结算账户ID
    
    pub amount: i64, // 收款金额
    
    pub discount_amount: Option<i64>, // 优惠金额
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub receipt_date: NaiveDateTime, // 收款日期
    
    pub payment_method: Option<String>, // 收款方式 (如 bank_transfer, cash, credit)
    
    pub receipt_status: i8, // 状态 (0=pending, 1=completed, 2=cancelled)
    
    pub remarks: Option<String>, // 备注
    
    pub department_code: String, // 部门编码
    
    pub department_id: i64, // 部门ID
    
    pub creator: Option<i64>, // 创建者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub create_time: NaiveDateTime, // 创建时间
    
    pub updater: Option<i64>, // 更新者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub update_time: NaiveDateTime, // 更新时间
    
}

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpReceiptPageResponse {
    
    pub id: i64, // 收款ID
    
    pub customer_id: i64, // 客户ID
    
    pub user_id: i64, // 关联用户ID
    
    pub settlement_account_id: Option<i64>, // 结算账户ID
    
    pub amount: i64, // 收款金额
    
    pub discount_amount: Option<i64>, // 优惠金额
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub receipt_date: NaiveDateTime, // 收款日期
    
    pub payment_method: Option<String>, // 收款方式 (如 bank_transfer, cash, credit)
    
    pub receipt_status: i8, // 状态 (0=pending, 1=completed, 2=cancelled)
    
    pub remarks: Option<String>, // 备注
    
    pub department_code: String, // 部门编码
    
    pub department_id: i64, // 部门ID
    
    pub creator: Option<i64>, // 创建者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub create_time: NaiveDateTime, // 创建时间
    
    pub updater: Option<i64>, // 更新者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub update_time: NaiveDateTime, // 更新时间

    /****************** 信息 ******************/

    pub customer_name: Option<String>, // 客户名

    pub settlement_account_name: Option<String>, // 结算账户名
    
}

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpReceiptBaseResponse {
    
    pub id: i64, // 收款ID
    
    pub customer_id: i64, // 客户ID
    
    pub user_id: i64, // 关联用户ID
    
    pub settlement_account_id: Option<i64>, // 结算账户ID
    
    pub amount: i64, // 收款金额
    
    pub discount_amount: Option<i64>, // 优惠金额
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub receipt_date: NaiveDateTime, // 收款日期
    
    pub payment_method: Option<String>, // 收款方式 (如 bank_transfer, cash, credit)
    
    pub receipt_status: i8, // 状态 (0=pending, 1=completed, 2=cancelled)
    
    pub remarks: Option<String>, // 备注

    pub details: Vec<ErpReceiptDetailBaseResponse>, // 收款列表

    pub attachments: Vec<ErpReceiptAttachmentBaseResponse>, // 收款附件列表
    
}

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpReceiptInfoResponse {
    
    pub id: i64, // 收款ID
    
    pub customer_id: i64, // 客户ID
    
    pub user_id: i64, // 关联用户ID
    
    pub settlement_account_id: Option<i64>, // 结算账户ID
    
    pub amount: i64, // 收款金额
    
    pub discount_amount: Option<i64>, // 优惠金额
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub receipt_date: NaiveDateTime, // 收款日期
    
    pub payment_method: Option<String>, // 收款方式 (如 bank_transfer, cash, credit)
    
    pub receipt_status: i8, // 状态 (0=pending, 1=completed, 2=cancelled)
    
    pub remarks: Option<String>, // 备注

    pub details: Vec<ErpReceiptDetailBaseResponse>, // 收款列表

    pub attachments: Vec<ErpReceiptAttachmentBaseResponse>, // 收款附件列表

    /****************** 信息 ******************/

    pub customer_name: Option<String>, // 客户名

    pub settlement_account_name: Option<String>, // 结算账户名
    
}